src/compiler/checker.ts

```diff
@@ -3422,12 +3471,18 @@ module ts {
                         else {
                             type = getDeclaredTypeOfSymbol(symbol);
                             if (type.flags & (TypeFlags.Class | TypeFlags.Interface) && type.flags & TypeFlags.Reference) {
-                                let typeParameters = (<InterfaceType>type).typeParameters;
-                                if (node.typeArguments && node.typeArguments.length === typeParameters.length) {
-                                    type = createTypeReference(<GenericType>type, map(node.typeArguments, getTypeFromTypeNode));
+                                // In a type reference, the outer type parameters of the referenced class or interface are automatically
+                                // supplied as type arguments and the type reference only specifies arguments for the local type parameters

```

Comment:
We would then need `instantiateType` to instantiate all local type declarations (because they potentially reference outer type parameters) which it isn't prepared to do. Also, if those local type declarations themselves have type parameters, we'd end up with type references where both the target and the type arguments need to be instantiated. These are all "landmines" for runaway generic instantiation that we could never detect. Right now, we can only detect runaway instantiations that originate in type references and only by making the outer type parameters be type parameters of the class/interface do we see all of them in one place.